---
title: "CNV calling report"
output: html_document
params:
    script:
    config:
    variants:
    bed:
    target:
        prefix: ""
        name: ""
    ref:
        prefix: ""
        name: ""
    date: "`r format(Sys.time(), '%d-%B-%Y')`"
---

```{r, include = FALSE}
source(params$script)

params <- list(config = "/groups/umcg-fg/tmp01/projects/pgx-passport/tools/asterix/src/main/python/cnvcaller/conf/config.yml",
               variants = "/groups/umcg-fg/tmp01/projects/pgx-passport/results/gsamd-24v3/analyses/select_corrective_variants/out.locus.bed",
               source = list(
                 prefix = "out"))
```

# Input

```{r, echo = FALSE}
# Load config
config <- load_yaml_config(params$config)
cnv_path_prefix <- params$source$prefix

# Curated variants
# variants_curated_bed <- read_tsv(
#   "/groups/umcg-wijmenga/tmp01/projects/pgx-pipeline/tools/PGx-passport-pilot/data/cyp2d6/configuration_bed_files/variants_curated_beta.bed",
#   col_names=c("Chrom", "Start", "End", "Name"))

# CYP2D6 bed file
cyp2d6loc <- read_tsv(config$base$bed,
                      col_names = c("Chrom", "Start", "End", "Name"))

# CYP2D6 bed file
cyp2d6bb <- read_tsv("/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-passport-pilot/data/cyp2d6/crucial_backbone.bed",
                      col_names = c("Chrom", "Start", "End", "Name"))

features <- read_tsv("/groups/umcg-fg/tmp01/projects/pgx-passport/tools/PGx-passport-pilot/data/cyp2d6/configuration_bed_files/exons_introns_ext.bed",
                    col_names = c("Chrom", "Start", "End", "Name")) %>%
  mutate(Feature = if_else(row_number() %% 2 == 0, "Exon", "Intron")) %>%
  group_by(Feature) %>%
  mutate(FeatureNumber = row_number(),
         Exon = if_else(Feature == "Exon", FeatureNumber, NA_integer_),
         Intron = if_else(Feature == "Intron", FeatureNumber, NA_integer_)) %>%
  ungroup() %>%
  rowwise() %>%
  mutate(FeatureStart = min(Start, End),
         FeatureEnd = max(Start, End)) %>%
  select(FeatureStart, FeatureEnd, Exon, Intron)

# Load variant selection
variants <- read_tsv(params$variants,
  col_names = c("Chrom", "Start", "End", "Name")) %>%
  mutate(Name = ordered(Name, levels = .$Name[order(.$Start)])) %>%
  rowwise() %>%
  mutate(StartDistance = min(0, Start - cyp2d6loc[1, "Start", drop = T]),
         EndDistance = max(0, End - cyp2d6loc[1, "End", drop = T]),
         Distance = c(StartDistance, EndDistance)[which.max(abs(c(StartDistance, EndDistance)))]) %>%
  ungroup() %>%
  group_by(Start, End, Chrom, Distance) %>%
  summarise(n = n()) %>%
  left_join(cyp2d6bb %>% select(-Name) %>% mutate(Crucial = T), by = c("Start", "End", "Chrom"))


  # Load variant selection
variants_ugli <- read_tsv(ugli_locus,
                     col_names = c("Chrom", "Start", "End", "Name")) %>%
  mutate(Name = ordered(Name, levels = .$Name[order(.$Start)])) %>%
  rowwise() %>%
  mutate(StartDistance = min(0, Start - cyp2d6loc[1, "Start", drop = T]),
         EndDistance = max(0, End - cyp2d6loc[1, "End", drop = T]),
         Distance = c(StartDistance, EndDistance)[which.max(abs(c(StartDistance, EndDistance)))]) %>%
  ungroup() %>%
  group_by(Start, End, Chrom, Distance) %>%
  summarise(n = n()) %>%
  left_join(cyp2d6bb %>% select(-Name) %>% mutate(Crucial = T), by = c("Start", "End", "Chrom"))

variants_plot <- variants_ugli %>%
  filter(Distance == 0) %>%
  ggplot(aes(x = Start, y = 0, yend = n, xend = Start, colour = !is.na(Crucial))) +
  geom_segment() +
  ylab("Number of assays for variant") +
  xlab("Genomic position")


plot_to_pdf(variants_plot, file.path(".", sprintf(
  "out.variants_ugli.%s.pdf", format(Sys.Date(), "%Y%m%d"))), width = 6, height = 3)



# The corrected intensities and projected intensities for the variants of interest
intensities_corrected <- read_intensities_processed(
  paste(cnv_path_prefix, "intensity_correction", "corrected", sep = "."), values_to = "R_corrected")

# The corrected intensities dataframe
corrected_intensities_data_frame <- read_intensities_raw(
  cnv_path_prefix) %>%
  inner_join(intensities_corrected, by = c("Sample_ID", "variant")) %>%
  mutate(theta = atan(Y/X),
         theta_ = atan(sqrt(Y)/sqrt(X)),
         X_corrected = X * (R_corrected / R),
         Y_corrected = Y * (R_corrected / R))

```

# Intensity correction

```{r, echo = FALSE}
intensities_per_type <- corrected_intensities_data_frame %>%
  pivot_longer(
    cols = c(X, Y, X_corrected, Y_corrected, R, R_corrected),
    names_pattern = "(.)(_corrected)?",
    names_to = c(".value", "corrected")) %>%
  dplyr::select(variant, Sample_ID, corrected, X, Y, R, theta) %>%
  mutate(corrected = ordered(
    if_else(corrected == "_corrected", "corrected", "uncorrected"),
    levels = c("uncorrected", "corrected"))) %>%
  inner_join(variants, by = c("variant" = "Name")) %>%
  mutate(variant = ordered(variant, levels = levels(variants$Name)))

intensities_per_type_sampled <- intensities_per_type %>%
  filter(Start %in% c(42522601, 42522613, 42523805, 42523843,
                      42523858, 42525044, 42525176, 42525772, 42526712))

scales_x <- intensities_per_type_sampled %>%
  group_by(variant) %>%
  group_map(~ scale_x_continuous(limits = c(
      min(unlist(.x[c("X", "Y")]), na.rm = TRUE),
      max(unlist(.x[c("X", "Y")]), na.rm = TRUE)))) %>%
  setNames(unique(sort(intensities_per_type_sampled$variant)))

scales_y <- intensities_per_type_sampled %>%
  group_by(variant) %>%
  group_map(~ scale_y_continuous(limits = c(
    min(unlist(.x[c("X", "Y")]), na.rm = TRUE),
    max(unlist(.x[c("X", "Y")]), na.rm = TRUE)))) %>%
  setNames(unique(sort(intensities_per_type_sampled$variant)))

# Create plot
correction_plot <- ggplot(intensities_per_type_sampled, aes(x=X, y=Y)) +
  rasterize(geom_point(
    shape = 16, alpha = 0.16, size = 0.2,
    aes(colour = R)), dpi = 300) +
  scale_colour_viridis_c(option = "viridis",
                         name = "R", na.value = "grey50") +
  facet_wrap(~ variant + corrected, scales = "free", ncol = 8) +
  facetted_pos_scales(y = rep(scales_y, each=2), x = rep(scales_x, each=2)) +
  theme(legend.position = "bottom",
        axis.text = element_blank(),
        aspect.ratio = 1)

plot_to_pdf(correction_plot, file.path(".", sprintf(
  "out.correction.%s.pdf", format(Sys.Date(), "%Y%m%d"))), width = 8, height = 11)

```
# Overview of variants
```{r, echo = FALSE}
# intensities_per_type_corrected <-
variants_of_interest <- intensities_per_type %>%
  filter(corrected == "corrected") %>%
  distinct(variant, Start) %>%
  cross_join(features) %>%
  filter(Start >= FeatureStart, Start <= FeatureEnd) %>%
  group_by(Start) %>%
  mutate(AssayCount = n_distinct(variant)) %>%
  filter(AssayCount >= 3 | Start == 42524244)

intensities_per_type_corrected <- intensities_per_type %>%
  filter(corrected == "corrected") %>%
  semi_join(variants_of_interest, by = "Start") %>%
  group_by(Start) %>%
  group_split()

for (intensities_grouped in intensities_per_type_corrected) {
  print(intensities_grouped)

  scales_x <- intensities_grouped %>%
  group_by(variant) %>%
  group_map(~ scale_x_continuous(limits = c(
    min(unlist(.x[c("X", "Y")]), na.rm = TRUE),
    max(unlist(.x[c("X", "Y")]), na.rm = TRUE)))) %>%
  setNames(unique(sort(intensities_grouped$variant)))

  scales_y <- intensities_grouped %>%
    group_by(variant) %>%
    group_map(~ scale_y_continuous(limits = c(
      min(unlist(.x[c("X", "Y")]), na.rm = TRUE),
      max(unlist(.x[c("X", "Y")]), na.rm = TRUE)))) %>%
    setNames(unique(sort(intensities_grouped$variant)))

  # Create plot
  correction_plot <- ggplot(intensities_grouped, aes(x=X, y=Y)) +
    rasterize(geom_point(
      shape = 16, alpha = 0.16, size = 0.2,
      aes(colour = R)), dpi = 300) +
    scale_colour_viridis_c(option = "viridis",
                           name = "R", na.value = "grey50") +
    facet_wrap2(~ variant, scales = "free", ncol = 6, nrow = 3) +
    facetted_pos_scales(y = scales_y, x = scales_x) +
    theme(legend.position = "bottom",
          axis.text = element_blank(), strip.text.x = element_text(size = 6),
          axis.ticks = element_blank(),
          aspect.ratio = 1)

  plot_to_pdf(correction_plot, file.path(".", sprintf(
    "out.corrected.%s.%s.pdf", head(intensities_grouped$Start, 1), format(Sys.Date(), "%Y%m%d"))), width = 6, height = 6)

}
```

```{r, echo = FALSE}
naive <- c(42523777)
curated <- c(42527902, 42526712, 42525782, 42525774, 42525616, 42525509, 42525194, 42525044, 42524796, 42523858, 42523855, 42523805, 42523777, 42523763, 42522630, 42525086)
curated_set <- variants_of_interest %>% filter(Start %in% curated)
curated_gsa_v3 <- curated_set %>% mutate(Chrom = 22, End = Start) %>% rename(Name = variant) %>% select(Chrom, Start, End, Name)
write.table(curated_gsa_v3, "curated_gsa_v3.bed", sep="\t", row.names = F, col.names = F, quote=F)
```

```{r, echo = FALSE}
naive_probabilities <- read_csv("out.naive_clustering.updated_probabilities.csv.gz") %>%
    rename("Sample_ID" = `Sample ID`) %>%
    rowwise() %>%
    mutate(
      Probabilities_Sum = rowSums(across(-Sample_ID, ~ exp(.x))),
      across(c(-Sample_ID, -Probabilities_Sum), ~ exp(.x) / Probabilities_Sum),
      Dosage = rowSums(across(c(-Sample_ID, -Probabilities_Sum), ~ ((as.numeric(cur_column()) + 2) * .x))),
    Probability_Max = max(c_across(c(-Sample_ID, -Dosage, -Probabilities_Sum))),
    Probability_Low = Probability_Max < 0.97,
    Hard_Dosage = round(Dosage)) %>% ungroup()

intensities_naive_clustering <- intensities_per_type %>%
  filter(corrected == "corrected") %>%
  semi_join(curated_gsa_v3, by=c("Start"="Start", "End"="End", "variant"="Name")) %>%
  inner_join(naive_probabilities, by = "Sample_ID")


for (intensities_grouped in intensities_naive_clustering %>%
  group_by(Start) %>%
  group_split()) {
  print(intensities_grouped)

  # Create plot
  correction_plot <- ggplot(intensities_grouped, aes(x=X, y=Y)) +
    rasterize(geom_point(
      shape = 16, alpha = 0.16, size = 0.2,
      aes(colour = Dosage)), dpi = 300) +
    scale_colour_viridis_c(option = "viridis",
                           name = "R", na.value = "grey50") +
    facet_wrap2(~ variant, ncol = 4, nrow = 4, trim_blank = FALSE) +
    coord_fixed() +
    theme(legend.position = "bottom", strip.text.x = element_text(size = 6), ,
          aspect.ratio = 1)

  plot_to_pdf(correction_plot, file.path(".", sprintf(
    "out.naive.%s.%s.pdf", head(intensities_grouped$Start, 1), format(Sys.Date(), "%Y%m%d"))), width = 6, height = 6)

}
```

```{r, echo = FALSE}
sample_scores <- read_csv(
  "out.variant_scores.fitted.csv.gz") %>%
  mutate(Pass = score > 0)

probabilities_processed <- read_csv(
  "out.cnv_probabilities.fitted.csv.gz") %>%
  rename("Sample_ID" = "Sample ID") %>%
  mutate(Cnv = A + B, Probability_Log=log(Probability + .Machine$double.eps)) %>%
  inner_join(sample_scores, by = c("Sample_ID", "Variant")) %>%
  filter(Pass)

cnv_per_variant_probabilities <- probabilities_processed %>%
  group_by(Sample_ID, Variant, Cnv) %>%
  summarise(Cnv_Probability = sum(Probability)) %>%
  group_by(Sample_ID, Variant) %>%
  filter(any(Cnv_Probability > 0.95))

probabilities_raw <- cnv_per_variant_probabilities %>%
  group_by(Sample_ID, Variant) %>%
  slice_max(Cnv_Probability) %>%
  inner_join(curated_gsa_v3, by = c("Variant" = "Name"))

cnv_per_variant_dosages <- probabilities_raw %>%
  inner_join(cnv_probabilities %>% group_by(Sample_ID) %>% slice_max(Cnv_Probability_Adjusted), by = c("Sample_ID"), suffix = c("_Overall", "_Variant")) %>%
  arrange(Start) %>%
  mutate(Cnv_Difference = Cnv_Variant - Cnv_Overall)

cnv_probabilities <- cnv_per_variant_probabilities %>%
  group_by(Sample_ID, Cnv) %>%
  summarise(
    Cnv_Probabilities_Summed = log(sum(Cnv_Probability, na.rm=T))) %>%
  group_by(Sample_ID) %>%
  mutate(
    Cnv_Probability_Total = log(sum(exp(Cnv_Probabilities_Summed), na.rm=T)),
    Cnv_Probability_Adjusted = exp(Cnv_Probabilities_Summed - Cnv_Probability_Total))

probabilities_adjusted <- cnv_probabilities %>%
  ungroup() %>%
  inner_join(
    probabilities_processed,
    by = c("Sample_ID", "Cnv"), multiple = "all") %>%
  group_by(Sample_ID, Variant, Cnv) %>%
  mutate(
    Cnv_Variant_Probability = sum(Probability, na.rm=T),
    Adjusted_Probability = Probability / Cnv_Variant_Probability * Cnv_Probability_Adjusted) %>%
  group_by(Sample_ID, Variant) %>%
  mutate(Max_Cnv_Variant_Probability = max(Cnv_Variant_Probability, na.rm=T),
         Cnv_Unadjusted = head(Cnv[Max_Cnv_Variant_Probability == Cnv_Variant_Probability], 1))

complement <- c("A"="T", "T"="A", "G"="C", "C"="G")

manifest <- read_csv("~/pgx-passport/data/reference/arrays/GSAMD-24v3-0-EA_20034606_A1.csv", skip = 7)
manifest_updated <- manifest %>%
  extract(col = SNP, into = c("A", "B"), regex = "\\[(\\w+)/(\\w+)\\]") %>%
  extract(col = IlmnID,
          into = c('IlmnID_Name', 'IlmnID_Strand','IlmnID_FwdRev', 'IlmnID_ProbeID'),
          regex = "(.+)_(T|B|P|M)_(F|R)_(\\d+)") %>%
  mutate(
    A_Fwd = case_when(
      IlmnID_FwdRev == "F" ~ A,
      IlmnID_FwdRev == "R" ~ complement[A]),
    B_Fwd = case_when(
      IlmnID_FwdRev == "F" ~ B,
      IlmnID_FwdRev == "R" ~ complement[B]))

write_delim(
  probabilities_adjusted_corrected %>%
  distinct(Sample_ID, Cnv, Cnv_Probability_Adjusted),
  "cnv_status.txt", delim=" ")

b_dosage <- probabilities_adjusted_corrected %>%
  filter(Cnv <= 2, Cnv_Probability_Adjusted > 0.95) %>%
  group_by(Sample_ID, Variant) %>%
  mutate(B_normalized = as.integer(B/Cnv * 2)) %>%
  arrange(B_normalized) %>%
  pivot_wider(
    id_cols=c(Variant, Sample_ID, Cnv, Cnv_Probability_Adjusted),
    values_from = "Adjusted_Probability",
    names_from = "B_normalized",
    names_prefix = "Genotypes_",
    values_fill = 0) %>%
  mutate(
    across(starts_with("Genotypes_"), ~ round(.x, digits = 3)),
    Total_Probobability = rowSums(across(starts_with("Genotypes_"))),
    across(starts_with("Genotypes_"), ~ .x / Total_Probobability)) %>%
  select(-one_of("Genotypes_NA")) %>%
  inner_join(manifest_updated, by = c("Variant"="Name")) %>%
  mutate(RefSnp_ID = Variant) %>%
  arrange(Sample_ID) %>%
  pivot_wider(
    id_cols=c(Chr, Variant, RefSnp_ID, MapInfo, A_Fwd, B_Fwd),
    names_from="Sample_ID",
    names_glue = "{.value}_Sample_{Sample_ID}",
    values_fill=0,
    values_from=starts_with("Genotypes_"))

ordering <- tibble(column_names = colnames(
  b_dosage %>% ungroup() %>% select(starts_with("Genotypes_")))) %>%
  separate(col=column_names, into=c("Genotype", "Sample"), sep="_Sample_", remove = FALSE) %>%
  arrange(Sample, match(Genotype, c("Genotype_0", "Genotype_1", "Genotype_2")))

write_delim(b_dosage %>% select(c(Chr, Variant, RefSnp_ID, MapInfo, A_Fwd, B_Fwd), ordering$column_names) %>% arrange(MapInfo), "genotype_probabilities_incl_deletions_2023-03-23.gen", delim=" ", col_names = F)

samples <- ordering %>%
  mutate(missing=0, sex='0') %>%
  select(c("ID_1"=Sample, "ID_2"=Sample), c(missing, sex)) %>%
  distinct()
write_delim(samples, "genotype_probabilities_incl_deletions_2023-03-23.sample", delim=" ")

predicted_genotypes <- probabilities_adjusted %>%
  mutate(Genotype = paste0(
    case_when(A == 0 & B == 0 ~ "0", TRUE ~ ""),
    case_when(A == 0 ~ "", A == 1 ~ "A", A == 2 ~ "AA", A > 2 ~ paste0(as.character(A), "A")),
    case_when(B == 0 ~ "", B == 1 ~ "B", B == 2 ~ "BB", B > 2 ~ paste0(as.character(B), "B"))),
         CNV = A + B,
         Genotype_Theta = log2(A / B)) %>%
  rename("A_Dosage" = "A", "B_Dosage" = "B")

fitted_intensities <- intensities_per_type %>%
  filter(corrected == "corrected") %>%
  inner_join(probabilities_raw, by=c("Sample_ID"="Sample_ID", "Start"="Start")) %>%
  group_by(Start) %>%
  arrange(desc(Probability)) %>%
  group_split()

for (intensities_grouped in fitted_intensities) {
  print(intensities_grouped)

  # Create plot
  fit_plot <- ggplot(intensities_grouped, aes(x=X, y=Y)) +
    rasterize(geom_point(
      alpha = 0.2, size = 1, shape = 21,
      aes(fill = Cnv, stroke = Probability < 0.95)), dpi = 300) +
    scale_discrete_manual(aesthetics = "stroke", values=c(0, 0.5)) +
    scale_fill_viridis_c(option = "viridis",
                           name = "Dosage", na.value = "grey50") +
    facet_wrap2(~ variant, ncol = 4, nrow = 4, trim_blank = FALSE) +
    coord_fixed() +
    theme(legend.position = "bottom", strip.text.x = element_text(size = 6), ,
          aspect.ratio = 1)

  plot_to_pdf(fit_plot, file.path(".", sprintf(
    "out.fit.%s.%s.pdf", head(intensities_grouped$Start, 1), format(Sys.Date(), "%Y%m%d"))), width = 8, height = 8)

}

probabilities_plot <- cnv_probabilities %>%
  slice_max(Cnv_Probability_Adjusted) %>%
  ggplot(aes(Cnv_Probability_Adjusted)) +
        geom_histogram(bins=30)

plot_to_pdf(probabilities_plot, file.path(".", sprintf(
  "out.probabilities.%s.pdf", format(Sys.Date(), "%Y%m%d"))), width = 4, height = 4)

```